### ISSUE TITLE
Consider using pyugrid!

### ISSUE THREAD

**Issue by ChrisBarker-NOAA** [NESII/python-ugrid#12]
Hi,

Just happened on your little project -- this is functionality we would like to have in pyugrid:

https://github.com/pyugrid/pyugrid

and pyugrid provides functionality that will make it a lot easier -- it alredy writes ugrid+CF-compatible netcdf files.

We're doing some re-structuring, but there will be a pyugrid.io subpackage where this could nicely live.

Come join us!

-Chris

**Comment by bekozi** [comment]
Hi, Chris. Sorry for not getting back to you sooner. It definitely makes sense to merge efforts. This package is standalone at the moment, and I do not expect to be able to rewrite it using a new API anytime soon. This package only works with flexible meshes, but it is likely portable to the triangular mesh topology. The goal is to slowly expand IO capabilities to work with other meshes. This package is used by OpenClimateGIS mainly. If this code can stay more or less intact but be housed in pyugrid, then there shouldn't be any problems! Could you point me to an an example of writing a CF-UGRID file using pyugrid? Thanks!

**Comment by bekozi** [comment]
Hi, Chris! Has there been progress in the creation of pyugrid's io package? In the meantime, I renamed this package to "ocgis-ugrid" to avoid confusion...

**Comment by ChrisBarker-NOAA** [comment]
No, but a start of the functionality is there -- we just need someone to refactor the code...

**Comment by ChrisBarker-NOAA** [comment]
A bit more:

The reading code is already in a separate module:
https://github.com/pyugrid/pyugrid/blob/master/pyugrid/read_netcdf.py

But the writting code is in the main module:

https://github.com/pyugrid/pyugrid/blob/master/pyugrid/ugrid.py

(not for any good reason :-) )

**Comment by bekozi** [comment]
Got it! I don't think it would be too hard to adapt the 2d flexible mesh code to fit into this paradigm. Our current code is being refactored for speed improvements when converting large datasets anyway... Let me take a closer look at everything and get back to you with some thoughts on how to move forward. I'll plan to contact you by email. Many thanks!

**Comment by ChrisBarker-NOAA** [comment]
Sounds good -- nice to share effort where we can.

**Comment by ocefpaf** [comment]
@bekozi before you read this note that I might be completely wrong about what I recommend below.  I don't really know the `ocgis-ugrid` and if it makes sense to have `pyugrid` as its dependency. With that said I would start here:

https://github.com/NESII/ocgis-ugrid/blob/master/src/ugrid/convert.py#L17-L21

if you read the netCDF file with `pyugrid` you will get unambiguous access to the nodes without hard-wiring the variable names, as the current code does, not to mention access to the other UGRID defined properties (center, faces, etc).  That would get `ocgis-ugrid` working with any UGRID compliant dataset.

Another place would be:

https://github.com/NESII/ocgis-ugrid/blob/master/src/ugrid/helpers.py#L283

the netcdf writing should be simpler if you use the `pyugrid.UGrid` class to construct the object, and then use the `.save_as_netcdf()` method to write the file.  The latter should take care of the UGRID convention bolierplate for you [[citation needed]](https://xkcd.com/285/) .

I know you are extremely busy, but if you think those are valid suggestions and want to give it a try let me know.  I am interested into seeing real cases where `pyugrid` is used for UGRID "conversations."

**Comment by bekozi** [comment]
@ocefpaf, this is woefully out-of-date code that hasn't been touched since I started working with ``pyugrid``. If you want to get a sense of where I am with the ``pyugrid`` fork, check out: https://github.com/NESII/pyugrid/tree/nfie. This has merged ``pyugrid/master`` frequently but may be a bit behind with @ChrisBarker-NOAA changes. I want to delete this ``ocgis-ugrid`` repository soon as the code has been ported to ``pyugrid``. The branch code is a work-in-progress that needs to be rebased and commented before attempting a PR...

I put up an example of how the ``FlexibleMesh`` object is being used here: https://gist.github.com/bekozi/8a232f509f1f80976f3a#file-prep_shapefiles-py-L30. The conversion is fully parallel except for the write which occurs on a single processor.

**Comment by ocefpaf** [comment]
You are subclassing `UGrid`. Nice!  (And much fancier than what I suggested :wink:)
